# HG changeset patch
# User Rik <rik@octave.org>
# Date 1360373834 28800
#      Fri Feb 08 17:37:14 2013 -0800
# Node ID d8f216d241cf5e319eb842ba4e2489215daa68d5
# Parent  26cba49d764169c40551d49ba242343cbba975d5
Fix off-by-1 error in run_history when called with no arguments.

* libinterp/interpfcn/oct-hist.cc(mk_tmp_hist_file): Subtract
2 from hist_count to account for last command and zero-based indexing.

diff --git a/libinterp/interpfcn/oct-hist.cc b/libinterp/interpfcn/oct-hist.cc
--- a/libinterp/interpfcn/oct-hist.cc
+++ b/libinterp/interpfcn/oct-hist.cc
@@ -369,33 +369,31 @@ get_int_arg (const octave_value& arg, in
 static std::string
 mk_tmp_hist_file (const octave_value_list& args,
                   bool insert_curr, const char *warn_for)
 {
   std::string retval;
 
   string_vector hlist = command_history::list ();
 
-  int hist_count = hlist.length ();
+  int hist_count = hlist.length () - 1;  // switch to zero-based indexing
 
   // The current command line is already part of the history list by
   // the time we get to this point.  Delete it from the list.
 
-  hist_count -= 2;
-
   if (! insert_curr)
     command_history::remove (hist_count);
 
-  hist_count--;
+  hist_count--;  // skip last entry in history list
 
   // If no numbers have been specified, the default is to edit the
   // last command in the history list.
 
+  int hist_beg = hist_count;
   int hist_end = hist_count;
-  int hist_beg = hist_count;
 
   bool reverse = false;
 
   // Process options.
 
   int nargin = args.length ();
 
   bool usage_error = false;
@@ -431,19 +429,17 @@ mk_tmp_hist_file (const octave_value_lis
   if (usage_error)
     {
       usage ("%s [first] [last]", warn_for);
       return retval;
     }
 
   if (hist_end < hist_beg)
     {
-      int t = hist_end;
-      hist_end = hist_beg;
-      hist_beg = t;
+      std::swap (hist_end, hist_beg);
       reverse = true;
     }
 
   std::string name = octave_tempnam ("", "oct-");
 
   std::fstream file (name.c_str (), std::ios::out);
 
   if (! file)
